libs/metagen/src/mdk_python/mod.rs

```diff
@@ -0,0 +1,338 @@
+    #[serde(flatten)]
+    #[garde(dive)]
+    pub base: crate::config::MdkGeneratorConfigBase,
+}
+
+impl MdkPythonGenConfig {
+    pub fn from_json(json: serde_json::Value, workspace_path: &Path) -> anyhow::Result<Self> {
+        let mut config: mdk_python::MdkPythonGenConfig = serde_json::from_value(json)?;
+        config.base.path = workspace_path.join(config.base.path);
+        config.base.typegraph_path = config
+            .base
+            .typegraph_path
+            .as_ref()
+            .map(|path| workspace_path.join(path));
+        Ok(config)
+    }
+}
+
+pub struct Generator {
+    config: MdkPythonGenConfig,
+}
+
+impl Generator {
+    pub const INPUT_TG: &'static str = "tg_name";
+    pub fn new(config: MdkPythonGenConfig) -> Result<Self, garde::Report> {
+        use garde::Validate;
+        config.validate(&())?;
+        Ok(Self { config })
+    }
+}
+
+impl crate::Plugin for Generator {
+    fn bill_of_inputs(&self) -> HashMap<String, GeneratorInputOrder> {
+        [(
+            Self::INPUT_TG.to_string(),
+            if let Some(tg_name) = &self.config.base.typegraph_name {
+                GeneratorInputOrder::TypegraphFromTypegate {
+                    name: tg_name.clone(),
+                }
+            } else if let Some(tg_path) = &self.config.base.typegraph_path {
+                GeneratorInputOrder::TypegraphFromPath {
+                    path: tg_path.clone(),
+                    name: self.config.base.typegraph_name.clone(),
+                }
+            } else {
+                unreachable!()
+            },
+        )]
+        .into_iter()
+        .collect()
+    }
+
+    fn generate(
+        &self,
+        inputs: HashMap<String, GeneratorInputResolved>,
+    ) -> anyhow::Result<GeneratorOutput> {
+        // return Ok(GeneratorOutput(Default::default()))
+        let tg = match inputs
+            .get(Self::INPUT_TG)
+            .context("missing generator input")?
+        {
+            GeneratorInputResolved::TypegraphFromTypegate { raw } => raw,
+            GeneratorInputResolved::TypegraphFromPath { raw } => raw,
+        };
+        let mut mergeable_output: IndexMap<PathBuf, Vec<RequiredObjects>> = IndexMap::new();
+
+        let mut tera = tera::Tera::default();
+        tera.add_raw_template("main_template", include_str!("static/main.py.jinja"))?;
+        tera.add_raw_template("types_template", include_str!("static/types.py.jinja"))?;
+        tera.add_raw_template("struct_template", include_str!("static/struct.py.jinja"))?;
+
+        let stubbed_funs = filter_stubbed_funcs(tg, &["python".to_string()])?;
+        for fun in &stubbed_funs {
+            if fun.mat.data.get("mod").is_none() {
+                continue;
+            }
+            let (_, script_path) = get_module_infos(fun, tg)?;
+            let base_path = self.config.base.path.clone();
+            let entry_point_path = if base_path.as_os_str().to_string_lossy().trim() == "" {
```

Comment:
Sort of related with the above comment but yeah, type modules per module it seems. Will this mean that the Struct impl will be duplicated for each module? I see that the impl allows multiple handlers in a single module but still, maybe it might be better to be more opinionated. 

Something like, only generate a single module with all the types and the decorators. The final layout will be then:
```
gen_target_dir/
  |- add.py
  |- identity.py
  |- mdk.py
```

`add.py` and `identity.py` will then import from `.mdk`. What do you think?

Note: in the future, we will have more code than the Struct impl in there like the hostcall wrapper so this duplication will multiply

Reply:
I see your reasoning. The fact that you common generated items will be de-duped alleviates sort of my worries. I still generally prefer one generated file over multiple (or one directory that I can just put in a `.gitignore` or `rm` with abandon). I also don't think length of generated files is of great concern but yeah, if it's all de-duped, it looks good to me.

Reply:
> or one directory that I can just put in a .gitignore or rm with abandon
Actually, you get this simply by refering the same file for all materializers.